octoprint_mrbeam/service/profile/profile.py

```diff
@@ -0,0 +1,482 @@
+                continue
+
+            if not entry.is_file():
+                continue
+
+            identifier = entry.name[:-len(".profile")]
+            results[identifier] = entry.path
+        return results
+
+    def _load_from_path(self, path):
+        if not os.path.exists(path) or not os.path.isfile(path):
+            return None
+
+        import yaml
+        with open(path) as f:
+            profile = yaml.safe_load(f)
+
+        if profile is None or not isinstance(profile, dict):
+            raise InvalidProfileError("Profile is None or not a dictionary")
+
+        if self._migrate_profile(profile):
+            try:
+                self._save_to_path(path, profile, allow_overwrite=True)
+            except:
+                self._logger.exception(
+                    "Tried to save profile to {path} after migrating it while loading, ran into exception".format(
+                        path=path))
+
+        profile = self._ensure_valid_profile(profile)
+
+        if not profile:
+            self._logger.warn("Invalid profile: %s" % path)
+            raise InvalidProfileError()
+        return profile
+
+    def _save_to_path(self, path, profile, allow_overwrite=False):
+        validated_profile = self._ensure_valid_profile(profile)
+        if not validated_profile:
+            raise InvalidProfileError()
+
+        if os.path.exists(path) and not allow_overwrite:
+            raise SaveError("Profile %s already exists and not allowed to overwrite" % profile["id"])
+
+        from octoprint.util import atomic_write
+        import yaml
+        try:
+            with atomic_write(path, "wb", max_permissions=0o666) as f:
+                yaml.safe_dump(profile, f, default_flow_style=False, indent="  ", allow_unicode=True)
+        except Exception as e:
+            self._logger.exception("Error while trying to save profile %s" % profile["id"])
+            raise SaveError("Cannot save profile %s: %s" % (profile["id"], str(e)))
+
+    def _remove_from_path(self, path):
+        try:
+            os.remove(path)
+            return True
+        except:
+            return False
+
+    def _get_profile_path(self, identifier):
+        return os.path.join(self._folder, "%s.profile" % identifier)
+
+    def _sanitize(self, name):
+        if name is None:
+            return None
+
+        if "/" in name or "\\" in name:
+            raise ValueError("name must not contain / or \\")
+
+        import string
+        valid_chars = "-_.() {ascii}{digits}".format(ascii=string.ascii_letters, digits=string.digits)
+        sanitized_name = ''.join(c for c in name if c in valid_chars)
+        sanitized_name = sanitized_name.replace(" ", "_")
+        return sanitized_name
+
+    def _migrate_profile(self, profile):
+        """
+        Subclasses must override this method.
+        """
+        return False
+        # raise NotImplementedError("Subclasses must implement _migrate_profile.")
```

Comment:
In the call, let's discuss
- I would maybe make the names of the methods in the profile service more self-explanatory or add docstrings
- What exactly is "identifier", because it's different to the id, no?
- What do we need this commented out code for?